### DISCUSSION TITLE
Is `store`/`state` per-request or global?

### DISCUSSION THREAD

**Discussion by crishoj** [elysiajs/elysia#702] [Q&A] [answered]
From https://elysiajs.com/essential/context.html#store it is not completely clear whether `state` assignments are per-request or shared among all requests.

Case in point: Suppose I would like to time response times with something like this...

```ts
export const requestLogging = new Elysia()
    .state('requestStart', 0)
    .onRequest(({store}) => {
        store.requestStart = Bun.nanoseconds()
    })
    .onResponse({as: 'global'}, ({request, set, store}) => {
        const url = new URL(request.url)
        const duration = formatDuration(Bun.nanoseconds() - store.requestStart)
        requestLogger.debug(
            chalk.grey('HTTP'), request.method, `${url.pathname}${url.search}`,
            chalk.grey('→ HTTP'), set.status, chalk.gray(`[${duration}]`))
    })
```

What happens if two requests are overlapping and handled concurrently? Would the second request overwrite `requestStart` in the store?

**Comment by SaltyAom** [comment, marked as answer]
`state` and `decorate` is **global** as stated in https://elysiajs.com/essential/handler.html#state
> State is a global mutable object or state shared across the Elysia app.

`resolve` and `derive` is **per request** as stated in https://elysiajs.com/essential/handler.html#derive
> Because derive is assigned once a new request starts

**Reply by crishoj** [reply]
Thank you @SaltyAom, that's perfectly clear.
